### PR TITLE
[DO NOT MERGE] [WIP] multiobjective inspiration

### PR DESCRIPTION
**Pull request by teytaud**

### CODE REVIEW

File: models/eval/inspirational_generation.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 import os
 import json
 from nevergrad.optimization import optimizerlib
+from nevergrad.functions.multiobjective.core import MultiobjectiveFunction
```
Comment (jrapin) [inline models/eval/inspirational_generation.py]: from nevergrad.functions import MultiobjectiveFunction

File: models/eval/inspirational_generation.py [hunk 2]
```diff
@@ -136,7 +137,7 @@ def gradientDescentOnInput(model,
 
     if nevergrad not in [None, 'CMA', 'DE', 'PSO',
                          'TwoPointsDE', 'PortfolioDiscreteOnePlusOne',
-                         'DiscreteOnePlusOne', 'OnePlusOne']:
+                         'DiscreteOnePlusOne', 'OnePlusOne', 'random', 'loss-covering', 'hypervolume', 'domain-covering']:
```
Comment (jrapin) [inline models/eval/inspirational_generation.py]: this mixes algorithms and selection mode, I dont get it and the docstring is not up to date.
Also, other modes dont seem to be used anyway since it loops on all selection options

File: models/eval/inspirational_generation.py [hunk 3]
```diff
@@ -200,11 +201,16 @@ def gradientDescentOnInput(model,
     gradientDecay = 0.1
 
     nImages = input.size(0)
+    assert nImages == 1
```
Comment (jrapin) [inline models/eval/inspirational_generation.py]: why always 1?

File: models/eval/inspirational_generation.py [hunk 4]
```diff
@@ -260,26 +266,27 @@ def resetVar(newVal):
         if lambdaD > 0:
 
             loss = -lambdaD * model.netD(noiseOut)[:, 0]
-            sumLoss += loss
+            combinedLoss[2] += loss
 
             if not randomSearch:
                 loss.sum(dim=0).backward()
 
         if nevergrad:
```
Comment (jrapin) [inline models/eval/inspirational_generation.py]: not modified, but: to check if it's None, always do `if nevergrad is not None`

File: models/eval/inspirational_generation.py [hunk 5]
```diff
@@ -200,11 +201,16 @@ def gradientDescentOnInput(model,
     gradientDecay = 0.1
 
     nImages = input.size(0)
+    assert nImages == 1
+    # assert randomSearch
+    thelosses = [1., 3., 3.]  # These numbers should be discussed...
+    target = MultiobjectiveFunction(lambda x: thelosses, tuple(thelosses))
```
Comment (jrapin) [inline models/eval/inspirational_generation.py]: this is constant? maybe use lambda x:x? and then target(losses) instead of target.compute_aggregate_loss?

File: models/eval/inspirational_generation.py (hunk 3 above)
Comment (teytaud) [reply models/eval/inspirational_generation.py]: For not mixing the different multiobjective functions.
Anyway the original code, I believe, is not good for the case nImages > 1..

File: models/eval/inspirational_generation.py (hunk 5 above)
Comment (teytaud) [reply models/eval/inspirational_generation.py]: When thelosses will be updated, this function will return something different. Tested with:
y=[1]

def biz(x):
    print(y)
    return y

biz(2)
biz(2)
y=[3]
biz(2)

File: models/eval/inspirational_generation.py (hunk 1 above)
Comment (teytaud) [reply models/eval/inspirational_generation.py]: thx.

File: models/eval/inspirational_generation.py (hunk 2 above)
Comment (teytaud) [reply models/eval/inspirational_generation.py]: fixed, now there is an additional "moo", in which case we loop over various Pareto sampling modes.

**Comment by teytaud** [comment]
@brozi : 
for using the latest master, install Nevergrad with:
pip install git+https://github.com/facebookresearch/nevergrad@master#egg=nevergrad
I propose that we launch this soon together.

File: models/eval/inspirational_generation.py (hunk 5 above)
Comment (jrapin) [reply models/eval/inspirational_generation.py]: I would have though that overriding y would unlink the external y from the one used internally but indeed...
Still, this is an ugly hack you should reaaaaaaally avoid doing for both clarity and avoid extremely weird bugs

Comment (teytaud) [reply models/eval/inspirational_generation.py]: The point is that Nevergrad is not equipped for Ask&Tell in multiobjective mode. I did not find a better solution for now.

File: models/eval/inspirational_generation.py [hunk 6]
```diff
@@ -260,26 +266,27 @@ def resetVar(newVal):
         if lambdaD > 0:
 
             loss = -lambdaD * model.netD(noiseOut)[:, 0]
-            sumLoss += loss
+            combinedLoss[2] += loss
```
Comment (brozi) [inline models/eval/inspirational_generation.py]: you can put just =

File: models/eval/inspirational_generation.py [hunk 7]
```diff
@@ -260,26 +266,27 @@ def resetVar(newVal):
...
 
             if not randomSearch:
                 loss.sum(dim=0).backward()
 
         if nevergrad:
             for i in range(nImages):
-                optimizers[i].tell(inps[i], float(sumLoss[i]))
```
Comment (brozi) [inline models/eval/inspirational_generation.py]: you could have an if nevergrad == 'moo' to avoid breaking the code when optimizing hte mono objectif

File: models/eval/inspirational_generation.py [hunk 8]
```diff
@@ -260,26 +266,27 @@ def resetVar(newVal):
...
             optimNoise.step()
 
         if optimalLoss is None:
             optimalVector = deepcopy(varNoise)
-            optimalLoss = sumLoss
+            optimalLoss = combinedLoss
 
         else:
-            optimalVector = torch.where(sumLoss.view(-1, 1) < optimalLoss.view(-1, 1),
+            optimalVector = torch.where(combinedLoss.view(-1, 1) < optimalLoss.view(-1, 1),
```
Comment (brozi) [inline models/eval/inspirational_generation.py]: you probably want combinedLoss.sum(dim=0) and same thing for optimalLoss here.
Otherwise you're taking parts of varNoise and parts of optimalLoss but those parts are not related to the losses that are better.

File: models/eval/inspirational_generation.py [hunk 9]
```diff
@@ -260,26 +266,27 @@ def resetVar(newVal):
...
                                         varNoise, optimalVector).detach()
-            optimalLoss = torch.where(sumLoss < optimalLoss,
-                                      sumLoss, optimalLoss).detach()
+            optimalLoss = torch.where(combinedLoss < optimalLoss,
```
Comment (brozi) [inline models/eval/inspirational_generation.py]: same, I think you should also take .sum(dim=0)

File: models/eval/inspirational_generation.py [hunk 10]
```diff
@@ -294,23 +301,34 @@ def resetVar(newVal):
                         outPath)
 
             print(str(iter) + " : " + formatCommand.format(
-                *["{:10.6f}".format(sumLoss[i].item())
+                *["{:10.6f}".format(combinedLoss[i].item())
```
Comment (brozi) [inline models/eval/inspirational_generation.py]: combinedLoss[:,i]
